File: 03 09 25/nextpermutation.rb
class Solution
  def next_permutation(nums)
    i = nums.length - 2

    # Step 1: Find first decreasing element from the end
    while i >= 0 && nums[i] >= nums[i + 1]
      i -= 1
    end

    if i >= 0
      j = nums.length - 1
      # Step 2: Find element just larger than nums[i]
      while nums[j] <= nums[i]
        j -= 1
      end
      nums[i], nums[j] = nums[j], nums[i]
    end

    # Step 3: Reverse the tail
    left = i + 1
    right = nums.length - 1
    while left < right
      nums[left], nums[right] = nums[right], nums[left]
      left += 1
      right -= 1
    end
  end
end

File: 03 09 25/nextpermutation.java
import java.util.Arrays;

class Solution {
    public void nextPermutation(int[] nums) {
        int i = nums.length - 2;

        // Step 1: Find first decreasing element from the end
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        if (i >= 0) {
            int j = nums.length - 1;
            // Step 2: Find element just larger than nums[i]
            while (nums[j] <= nums[i]) {
                j--;
            }
            swap(nums, i, j);
        }

        // Step 3: Reverse the tail
        reverse(nums, i + 1, nums.length - 1);
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    private void reverse(int[] nums, int left, int right) {
        while (left < right) {
            swap(nums, left, right);
            left++;
            right--;
        }
    }

